x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/retriever/QueryRuleRetrieverBuilder.java

```diff
@@ -0,0 +1,211 @@
+            throw LicenseUtils.newComplianceException("Query Rules");
+        }
+        try {
+            return PARSER.apply(parser, context);
+        } catch (Exception e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    private final List<String> rulesetIds;
+    private final Map<String, Object> matchCriteria;
+
+    public QueryRuleRetrieverBuilder(
+        List<String> rulesetIds,
+        Map<String, Object> matchCriteria,
+        RetrieverBuilder retrieverBuilder,
+        int rankWindowSize
+    ) {
+        super(new ArrayList<>(), rankWindowSize);
+        this.rulesetIds = rulesetIds;
+        this.matchCriteria = matchCriteria;
+        addChild(new QueryRuleRetrieverBuilderWrapper(retrieverBuilder));
+    }
+
+    public QueryRuleRetrieverBuilder(
+        List<String> rulesetIds,
+        Map<String, Object> matchCriteria,
+        List<RetrieverSource> retrieverSource,
+        int rankWindowSize,
+        String retrieverName
+    ) {
+        super(retrieverSource, rankWindowSize);
+        this.rulesetIds = rulesetIds;
+        this.matchCriteria = matchCriteria;
+        this.retrieverName = retrieverName;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public boolean isAllowedAsChildRetriever() {
+        return false;
+    }
+
+    @Override
+    protected SearchSourceBuilder createSearchSourceBuilder(PointInTimeBuilder pit, RetrieverBuilder retrieverBuilder) {
+        var ret = super.createSearchSourceBuilder(pit, retrieverBuilder);
+        checkValidSort(ret.sorts());
+        ret.query(new RuleQueryBuilder(ret.query(), matchCriteria, rulesetIds));
+        return ret;
+    }
+
+    private static void checkValidSort(List<SortBuilder<?>> sortBuilders) {
+        if (sortBuilders.isEmpty()) {
+            return;
+        }
+
+        if (sortBuilders.getFirst() instanceof ScoreSortBuilder == false) {
+            throw new IllegalArgumentException("[" + NAME + "] retriever only supports sort by score, got: " + sortBuilders);
+        }
+    }
+
+    @Override
+    public void doToXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.array(RULESET_IDS_FIELD.getPreferredName(), rulesetIds.toArray());
+        builder.startObject(MATCH_CRITERIA_FIELD.getPreferredName());
+        builder.mapContents(matchCriteria);
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryRuleRetrieverBuilder clone(List<RetrieverSource> newChildRetrievers) {
+        return new QueryRuleRetrieverBuilder(rulesetIds, matchCriteria, newChildRetrievers, rankWindowSize, retrieverName);
+    }
+
+    @Override
+    protected RankDoc[] combineInnerRetrieverResults(List<ScoreDoc[]> rankResults) {
+        assert rankResults.size() == 1;
+        ScoreDoc[] scoreDocs = rankResults.getFirst();
+        RankDoc[] rankDocs = new RankDoc[scoreDocs.length];
```

Comment:
How would `explain` look for this retriever? Adding a class that extends `RankDoc` gives us the option to explain how a match was made, so could we maybe add some info on which (if any) rules matched against a hit ?

Reply:
I love the idea of adding matching rules! Unfortunately, the only ways that I can think to do this would require a second client call to evaluate the ruleset. The extra time required to do that would probably be OK for explain, but I don't see a good way to do that in retrievers today. 

I don't want to go based on score (if it has the huge constant score) because that could put us in a bind in the future if we add other rules - already, we're not showing matched `exclude` rules since the results won't be returned. 

I did add some information, though, so Explain now looks like this: 

```
"description": "query rules evaluated rules from rulesets [my-ruleset] and match criteria {bar=bar, foo=foo}",
```

If you have better ideas they're very welcome 😄